src/main/java/com/gmail/nossr50/listeners/WorldListener.java

```diff
@@ -29,9 +31,16 @@ public void onStructureGrow(StructureGrowEvent event) {
         if(WorldBlacklist.isWorldBlacklisted(event.getWorld()))
             return;
 
-        // Using 50 ms later as I do not know of a way to run one tick later (safely)
-        plugin.getFoliaLib().getImpl().runLater(() -> {
-            for (BlockState blockState : event.getBlocks()) {
+        // Under Folia, any two loaded and adjacent chunks will be in the same ticking region. To avoid scheduling many
+        // tasks, we will use this assumption.
+        // Therefore, we will schedule the task such that all blocks in the event are handled from the ticking region of
+        // the first block in the event.
+        // Without folia, this will run on the main tick.
+        List<BlockState> blocks = event.getBlocks();
+        BlockState referenceBlock = blocks.get(0);
+        plugin.getFoliaLib().getImpl().runAtLocationLater(referenceBlock.getLocation(), () -> {
+            for (BlockState blockState : blocks) {
+                // The HashChunkManager is thread-safe, we can safely call it from a non-single-threaded environment
```

Comment:
This is not true

Reply:
I will review this again, as stated above, I don't remember everything I had checked back then. It's been a while.

Reply:
I have no idea why I thought this was thread-safe. I do in-fact agree. However, simply changing that class to using concurrent maps could introduce side-effects that would go further than just folia. It's probably better to stick with a more risky folia impl than to risk messing up the stable releases under normal bukkit.

I can remove the comment or compeletely revert that change. However, this chunk manager is already being access unsafely by other event listeners under folia and therefore isn't synced to any particular thread under a Folia environment.

Reply:
The proper solution is most likely changing the HashChunkManager to be thread safe, however, that would be a complex task and require a lot of testing. Hearing that it's already being used unsafely makes sense and is unfortunate, this should be addressed, or Folia server owners will run into some problems.